From fad645a0ca0f438b62dc3a332b34388bf9fd9fd0 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 5 Jun 2024 13:52:06 +0800
Subject: [PATCH 81/81] HACK: glamor: Clear the created empty FBO

To avoid the initial dirty frame of Cheese APP.

Tested on RK3588 EVB with:
/etc/init.d/lightdm restart
sleep 5
timeout 5 glmark2-es2
while sleep 1; do timeout 3 cheese; done

Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 glamor/glamor.h     |  1 +
 glamor/glamor_fbo.c | 19 +++++++++++++++++--
 2 files changed, 18 insertions(+), 2 deletions(-)

diff --git a/glamor/glamor.h b/glamor/glamor.h
index 698f9b880..fcec3f09c 100644
--- a/glamor/glamor.h
+++ b/glamor/glamor.h
@@ -127,6 +127,7 @@ extern _X_EXPORT Bool glamor_destroy_pixmap(PixmapPtr pixmap);
 #define GLAMOR_CREATE_PIXMAP_CPU        0x100
 #define GLAMOR_CREATE_PIXMAP_FIXUP      0x101
 #define GLAMOR_CREATE_FBO_NO_FBO        0x103
+#define GLAMOR_CREATE_FBO_CLEAR_FBO     0x104
 #define GLAMOR_CREATE_NO_LARGE          0x105
 #define GLAMOR_CREATE_PIXMAP_NO_TEXTURE 0x106
 #define GLAMOR_CREATE_FORMAT_CBCR       0x107
diff --git a/glamor/glamor_fbo.c b/glamor/glamor_fbo.c
index 2f242d397..2ab527ae3 100644
--- a/glamor/glamor_fbo.c
+++ b/glamor/glamor_fbo.c
@@ -124,6 +124,9 @@ glamor_create_fbo_from_tex(glamor_screen_private *glamor_priv,
         if (glamor_pixmap_ensure_fb(glamor_priv, fbo) != 0) {
             glamor_destroy_fbo(glamor_priv, fbo);
             fbo = NULL;
+        } else if (flag == GLAMOR_CREATE_FBO_CLEAR_FBO) {
+            glClearColor(0.0, 0.0, 0.0, 1.0);
+            glClear(GL_COLOR_BUFFER_BIT);
         }
     }
 
@@ -173,6 +176,9 @@ glamor_create_fbo(glamor_screen_private *glamor_priv,
     if (!tex) /* Texture creation failed due to GL_OUT_OF_MEMORY */
         return NULL;
 
+    if (flag != GLAMOR_CREATE_FBO_NO_FBO)
+        flag = GLAMOR_CREATE_FBO_CLEAR_FBO;
+
     return glamor_create_fbo_from_tex(glamor_priv, pixmap, w, h,
                                       tex, flag);
 }
@@ -353,14 +359,23 @@ glamor_pixmap_ensure_fbo(PixmapPtr pixmap, int flag)
     }
     else {
         /* We do have a fbo, but it may lack of fb or tex. */
-        if (!pixmap_priv->fbo->tex)
+        if (!pixmap_priv->fbo->tex) {
             pixmap_priv->fbo->tex =
                 _glamor_create_tex(glamor_priv, pixmap, pixmap->drawable.width,
                                    pixmap->drawable.height);
+            if (flag != GLAMOR_CREATE_FBO_NO_FBO)
+                flag = GLAMOR_CREATE_FBO_CLEAR_FBO;
+        }
 
-        if (flag != GLAMOR_CREATE_FBO_NO_FBO && pixmap_priv->fbo->fb == 0)
+        if (flag != GLAMOR_CREATE_FBO_NO_FBO && pixmap_priv->fbo->fb == 0) {
             if (glamor_pixmap_ensure_fb(glamor_priv, pixmap_priv->fbo) != 0)
                 return FALSE;
+
+            if (flag == GLAMOR_CREATE_FBO_CLEAR_FBO) {
+                glClearColor(0.0, 0.0, 0.0, 1.0);
+                glClear(GL_COLOR_BUFFER_BIT);
+            }
+        }
     }
 
     return TRUE;
-- 
2.38.5

